datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs

```diff
@@ -1473,7 +1470,123 @@ impl<'a, S: SimplifyInfo> TreeNodeRewriter for Simplifier<'a, S> {
+                op: Operator::And,
+                right,
+            }) if are_inlist_and_eq_and_match_neg(
+                left.as_ref(),
+                right.as_ref(),
+                false,
+                false,
+            ) =>
+            {
+                match (*left, *right) {
+                    (Expr::InList(l1), Expr::InList(l2)) => {
+                        return inlist_intersection(l1, l2, false).map(Transformed::yes);
+                    }
+                    // Matched previously once
+                    _ => unreachable!(),
+                }
+            }
+
+            Expr::BinaryExpr(BinaryExpr {
+                left,
+                op: Operator::And,
+                right,
+            }) if are_inlist_and_eq_and_match_neg(
+                left.as_ref(),
+                right.as_ref(),
+                true,
+                true,
+            ) =>
+            {
+                match (*left, *right) {
+                    (Expr::InList(l1), Expr::InList(l2)) => {
+                        return inlist_union(l1, l2, true).map(Transformed::yes);
+                    }
+                    // Matched previously once
+                    _ => unreachable!(),
+                }
+            }
+
+            Expr::BinaryExpr(BinaryExpr {
+                left,
+                op: Operator::And,
+                right,
+            }) if are_inlist_and_eq_and_match_neg(
+                left.as_ref(),
+                right.as_ref(),
+                false,
+                true,
+            ) =>
+            {
+                match (*left, *right) {
+                    (Expr::InList(l1), Expr::InList(l2)) => {
+                        return inlist_except(l1, l2).map(Transformed::yes);
+                    }
+                    // Matched previously once
+                    _ => unreachable!(),
+                }
+            }
+
+            Expr::BinaryExpr(BinaryExpr {
+                left,
+                op: Operator::And,
+                right,
+            }) if are_inlist_and_eq_and_match_neg(
+                left.as_ref(),
+                right.as_ref(),
+                true,
+                false,
+            ) =>
+            {
+                match (*left, *right) {
+                    (Expr::InList(l1), Expr::InList(l2)) => {
+                        return inlist_except(l2, l1).map(Transformed::yes);
+                    }
+                    // Matched previously once
+                    _ => unreachable!(),
+                }
+            }
+
+            Expr::BinaryExpr(BinaryExpr {
+                left,
+                op: Operator::Or,
+                right,
+            }) if are_inlist_and_eq_and_match_neg(
+                left.as_ref(),
+                right.as_ref(),
+                true,
+                true,
+            ) =>
+            {
+                match (*left, *right) {
+                    (Expr::InList(l1), Expr::InList(l2)) => {
```

Comment:
Currently, those inlist set rules does not support column yet, so we don't need `as_list`.